bpf/bpf_sock.c

```diff
@@ -369,6 +369,12 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 	}
 #endif /* ENABLE_L7_LB */
 
+	/* The service count is set to number of active backends. */
+	if (svc->count == 0) {
+		update_metrics(0, METRIC_EGRESS, REASON_LB_NO_ACTIVE_BACKEND_SLOT);
+		return -ENOENT;
+	}
```

Comment:
Should we also return an error from connect() / sendmsg() ?

Reply:
That's already the case. It's a common function - https://github.com/cilium/cilium/blob/master/bpf/bpf_sock.c#L602.

Reply:
> You meant to ask if we should bubble up the error to the caller? I'm not sure if that's useful as the errors are cilium specific. Applications/kernel code can just ignore the errors. The cilium
error metrics are useful for cilium users though.

We can return `SYS_REJECT` (in addition to updating the metrics). Otherwise we rely on the per-packet LB to catch & drop such packets (https://github.com/cilium/cilium/pull/21539) - and the per-packet LB might not always be active.

But I'm happy to defer the `SYS_REJECT` for now, introducing better visibility is a good start.

Reply:
>the per-packet LB might not always be active

What do you mean by this? If service translation isn't performed in bpf_sock, then the packets will be dropped in bpf_lxc based on the #21539 changes.

Reply:
Regarding, `SYS_REJECT`, we have a precedent for it in `bpf_sock`, but the motivations are totally different. I'm not convinced that we need to block connections because there are no service backends. Not to mention this doesn't work for UDP as there are no `connect` calls. As mentioned offline, it's better to drop such packets in bpf_lxc.

Reply:
> What do you mean by this? If service translation isn't performed in bpf_sock, then the packets will be dropped in bpf_lxc

`bpf_lxc` doesn't always compile the per-packet LB in (see below). I don't see how we can fully rely on it to drop such packets.

https://github.com/cilium/cilium/blob/4bafe1750f7b67fcab4b057c6e1aa9e8b9142584/bpf/bpf_lxc.c#L57-L69

Reply:
> If service translation isn't performed in bpf_sock, then the packets will be dropped in bpf_lxc

How do we handle connections from the host netns in this case (where bpf_lxc is definitely not available) ?

Reply:
> Regarding, SYS_REJECT, we have a precedent for it in bpf_sock, but the motivations are totally different. I'm not convinced that we need to block connections because there are no service backends.

Ack, I'm not 100% convinced either yet that we should return `SYS_REJECT`. As stated previously, happy to defer that part for now. But I believe we do need to consider it as an option.

Reply:
>How do we handle connections from the host netns in this case (where bpf_lxc is definitely not available) ?

We can potentially add a blackhole route for service vips. That's out of the scope of this PR though.

Reply:
> This PR is only adding visibility checks.

Fair enough. Mind updating the PR description? Right now it claims to bring a "fix" for the issue of "datapath may end up selecting a non-active backend".